### PR TITLE
Feature/reworks-hardware-interfaces

### LINKED ISSUE #5: Consolidate V5 Motor config into single file
Adding and re-configuring the V5 port configuration should only occur in a single file to remove redundancy and reduce accidental conflict

### LINKED ISSUE #7: Develop Base Class for V5 Robot
Main Competition Loop is implemented through ghost_ros/robot_state_machine_node.cpp node. We have hard-coded swerve and autonomous routines in this file.

This should be converted to an abstract base interface so we can implement a variety of robots without having to repeat the structure which subscribes and publishes V5 Msgs and which selects between auton/teleop/disabled.

Bonus is to provide motor and encoder class so that we can abstract ROS Topics out.

### PR DESCRIPTION
**Pull request by MaxxWilson**
# PR Summary
[PR Link](https://github.com/VEXU-GHOST/VEXU_GHOST/pull/39)

### Description
Huge rework of hardware interfaces. This is basically a full version upgrade of the entire repository so that ALL information about the hardware (motor/sensor definitions and configuration) is runtime configurable and automatically synced across devices.

### Reviewers
Tag reviewers.

- Required:
  
- Optional:

---
### Issue / Motivation
Previously, the configuration of different devices on the robot (i.e. sensors, motors, joystick, etc) could only be modified at compile-time. This meant that to change or tune a PID gain, we would have to recompile the entire repository and we had a gnarly dependency structure (the equivalent of global variables across all our ROS packages). Nasty, but required because we need this info at compile-time for the VEX Brain.

### Reviewer Guide
Full disclosure, this PR is WAY too big. This is a bad example of an ideal PR :clown_face:. Try to focus on:
1. High-level Interfaces (Approach section)
- Do they make sense?
- Could another team use this with sufficient documentation?
- Are there missing interfaces or complaints we can address early?
2. Test Coverage
- Are there key classes missing coverage right now?
- Are there test suites which don't fully cover all edge-cases or functionality of their respective feature?

You should probably ignore the changelog, and most of the package changes outside of ghost_v5_interfaces and ghost_ros_interfaces. I don't look at individual commits, only the Files Changed section.

Run Tests:
`cd ~/VEXU_GHOST`
`./scripts/test_pkg.sh ghost_over_under ghost_v5_interfaces ghost_ros_interfaces`

### Approach
This PR reworks the core of the codebase around a few key base classes, DeviceConfig, DeviceData, DeviceConfigMap, and RobotHardwareInterface.

#### DeviceConfig and DeviceData
DeviceConfig and DeviceData are data structures which represent the parameters of a given device. Config values are constant after they are loaded at boot and Data values change at every control loop. These are both abstract / pure virtual, so we inherit from them for each device we use on the robot. This makes it easy to unit-test that individual devices are loaded, serialized/deserialized, and interfaced properly. There are separate utility functions to load each DeviceConfig from YAML and to convert DeviceData to/from ROS msgs included under ghost_v5_interfaces/util and ghost_ros_interfaces/msg_helpers, respectively.

image

DeviceBase provides a few useful base methods (mostly for casting to derived types). DevicePair is also included for utility when storing data in a larger data structure.
image

#### DeviceConfigMap
All the devices are loaded into a single DeviceConfigMap, which represents the different ports on the V5 Brain. DeviceConfigMap basically just wraps a map from device_name to DeviceConfig. It also maintains a map of port_numbers to ensure that duplicate ports are not loaded (its impossible to plug two sensors into one port).

image

#### RobotHardwareInterface (RHI)
The DeviceConfigMap object is then used to initialize a RobotHardwareInterface (RHI), which is where a user can set and retrieve motor/sensor data. Essentially, you pass a YAML into a ROS node, it instantiates an RHI, and then you can directly address the robot hardware without dealing with any communication protocols in between. The RHI should implement any interfaces we would expect to have on the VEX Brain.

image

#### Build Process
The crucial component to all this is that a DeviceConfigMap can be loaded at runtime from a YAML file **_or_** it can be exported to raw C code and included at compile-time. Regardless of how it is created, it can be passed into an RHI the same way, so we can sync all the code across both devices despite the runtime/compile-time constraints. This functionality is included in ghost_v5_interfaces/util/device_config_factory_utils and its associated test suite.

This means you can write ONE text file with all your motors/sensors and the build process takes care of the rest. It also allows future Gazebo simulations to share parameters with the hardware configuration directly (rather than maintaining a copy of the file across the two different devices).

image

#### Automatic
- ghost_v5_interfaces/test/util/test_device_config_factory_utils.cpp is a god-tier piece of machinery which loads a DeviceConfigMap from YAML, generates source code, compiles the source code, and loads it as a plugin to validate the full codegen pipeline for the VEX Brain.
- ghost_v5_interfaces/test/test_robot_hardware_interface.cpp is probably the next most important test suite. RobotHardwareInterface is the primary access point for addressing motors and sensors, so this class needs to be 1) very clear and easy to use, and 2) bulletproof.
- ghost_v5_interfaces/test/devices includes a test suite for each implemented device type and the base device classes
- ghost_ros_interfaces/test/* has two test files for converting devices and the RobotHardwareInterface to/from ROS msg
#### Hardware Testing
- Tested the full control pipeline
1. sensor data is loaded into RobotHardwareInterface (RHI) on the VEX Brain
2. RHI is serialized and sent over microUSB to laptop
3. RHI is deserialized and published to a ROS topic by the serial node
4. Data is received in the state machine, which has loaded an example controller plugin
5. User code sets a specific motor to a given speed based on joystick input
6. Command data is sent back over ROS topic, through the serial interface, processed on the brain, and results in motor motion
- Viewed various sensor/motor data via PlotJuggler in real-time to validate data is properly passed between Jetson and V5 Brain.

### Changelog
**ghost_v5_interfaces (new)**
- Adds base interfaces for hardware devices (sensors/motors), including DeviceBase, DeviceConfig, and DeviceData
- Adds data structure for interacting with device tree via DeviceConfigMap
- Defines devices for Joystick, Motor, and RotationSensor
- Adds utility functions to load devices and configs from YAML
- Adds utility functions to load a DeviceConfigMap from YAML and to generate source code from a DeviceConfigMap object.
- Adds RobotHardwareInterface class as primary abstraction for addressing motors and sensors from user code.

**ghost_ros_interfaces (new)**
- Adds ros node which loads user code at runtime via plugins. This handles ROS Interfacing and state machine transitions for competition (autonomous, teleop, disabled, etc.)
- Adds msg helpers to convert common Device types to/from ROS msg

**ghost_msgs**
- Changes msg types to account for new interfaces
- Adds a few currently unsupported PROS fields that can be implemented later (if desired) without breaking msg compatibility.

**ghost_control**
- Modifies motor_controller to take in separate config structs for Controller, Filter, and Model.
- Fixes clamp bug which prevented proper voltage output from control calculation
- Adds DCMotorModel constructor which takes in a Config struct

**ghost_estimation**
- Modifies LowPassFilters to take in a Config struct

**ghost_planners (new)**
- Migrates all CasADi, MPC, and Trajectory Optimization work to this package

**ghost_serial**
- Adds error handling when readMsgFromSerial is given a vector of the wrong size.
- Changes some interfaces from unsigned char[] to std::vector<unsigned char> for safety
- Adds error rejection for junk bytes at startup (PROS terminal header)
- Various cleanup and debug improvement
- Moves Device specific compile flags to global cmake config
- readds python script to print bytestream to terminal for debugging

**ghost_util (new)**
- Split off of ghost_common
- Migrated byte utils from ghost_serial
- Adds helpers to deal with byte packing on systems with different architectures
- Adds templated helped function to load different parameters from a YAML file

**ghost_pros**
- Adds ScreenInterface, which allows for thread-safe scrolling terminal output to the V5 Brain screen
- Adds error checking for runtime-errors in main loops which display to the V5 Brain screen

**ghost_ros**
- Moves to deprecated and begins migrating useful code out.

### Checklist
- [x] Confirmed all tests pass on a clean build (ghost_sim has build failures that will need to be addressed in a separate PR that migrates sim)
- [x] Added reviewers in Github
- [x] Posted PR Summary to Discord PR's Channel
- [x] Ran uncrustify on any modified C++ files
- [x] Ran Colcon Lint for any modified CMakeLists.txt or Package.xml (Not doing this by hand, we need it in CI Pipeline)

### CODE REVIEW

**Comment by JakeWendling** [comment]
11_Robots/ghost_over_under/CMakeLists.txt includes the include directory that does not exist and does not compile because of this

File: 11_Robots/ghost_over_under/package.xml [hunk 1]
```diff
@@ -1,24 +1,21 @@
...
-  <depend>ament_index_cpp</depend>
-  <depend>urdfdom</depend>
-  <depend>ghost_common</depend>
 
   <test_depend>ament_lint_auto</test_depend>
   <test_depend>ament_lint_common</test_depend>
 
+  <exec_depend>ghost_ros_interfaces</exec_depend>
+  <exec_depend>ghost_swerve</exec_depend>
+  <exec_depend>rviz</exec_depend>
```
Comment (karmanyaahm) [inline 11_Robots/ghost_over_under/package.xml]: This should be `rviz2`. There is no rviz in humble.
```suggestion
  <exec_depend>rviz2</exec_depend>
```

File: 11_Robots/ghost_over_under/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+ament_export_include_directories(${INCLUDE})
+
+####################
+##### Install ######
+####################
+install(
+  DIRECTORY include/
+  DESTINATION include
+)
```
Comment (JakeWendling) [inline 11_Robots/ghost_over_under/CMakeLists.txt]:
```suggestion
```

The include folder does not exist so this should be deleted out for the code to compile

Comment (MaxxWilson) [reply 11_Robots/ghost_over_under/CMakeLists.txt]: Go ahead and delete instead of comment out, and good to go

**Review by melissajcruz — APPROVED** [review]
lgtm!
